// inferno-server/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var renderToString_1 = require("./renderToString");
exports.renderToString = renderToString_1.default;
exports.renderToStaticMarkup = renderToString_1.renderToStaticMarkup;
var renderToString_stream_1 = require("./renderToString.stream");
exports.streamAsString = renderToString_stream_1.default;
exports.RenderStream = renderToString_stream_1.RenderStream;
exports.streamAsStaticMarkup = renderToString_stream_1.streamAsStaticMarkup;
var renderToString_queuestream_1 = require("./renderToString.queuestream");
exports.streamQueueAsString = renderToString_queuestream_1.default;
exports.RenderQueueStream = renderToString_queuestream_1.RenderQueueStream;
exports.streamQueueAsStaticMarkup = renderToString_queuestream_1.streamQueueAsStaticMarkup;
exports.default = {
    renderToString: renderToString_1.default,
    renderToStaticMarkup: renderToString_1.renderToStaticMarkup,
    streamAsString: renderToString_stream_1.default,
    streamAsStaticMarkup: renderToString_stream_1.streamAsStaticMarkup,
    RenderStream: renderToString_stream_1.RenderStream,
    RenderQueueStream: renderToString_queuestream_1.RenderQueueStream,
    streamQueueAsString: renderToString_queuestream_1.default,
    streamQueueAsStaticMarkup: renderToString_queuestream_1.streamQueueAsStaticMarkup
};
