// inferno-router/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Link_1 = require("./Link");
exports.Link = Link_1.default;
var IndexLink_1 = require("./IndexLink");
exports.IndexLink = IndexLink_1.default;
var Route_1 = require("./Route");
exports.Route = Route_1.default;
var IndexRoute_1 = require("./IndexRoute");
exports.IndexRoute = IndexRoute_1.default;
var Redirect_1 = require("./Redirect");
exports.Redirect = Redirect_1.default;
exports.IndexRedirect = Redirect_1.default;
var Router_1 = require("./Router");
exports.Router = Router_1.default;
var match_1 = require("./match");
exports.match = match_1.default;
var createRoutes_1 = require("./createRoutes");
exports.createRoutes = createRoutes_1.default;
var RouterContext_1 = require("./RouterContext");
exports.RouterContext = RouterContext_1.default;
exports.default = {
    Route: Route_1.default,
    IndexRoute: IndexRoute_1.default,
    Redirect: Redirect_1.default,
    IndexRedirect: Redirect_1.default,
    Router: Router_1.default,
    RouterContext: RouterContext_1.default,
    Link: Link_1.default,
    IndexLink: IndexLink_1.default,
    match: match_1.default,
    createRoutes: createRoutes_1.default
};
