// inferno-router/dist/RouterContext.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var inferno_component_1 = require("inferno-component");
var RouterContext = (function (_super) {
    __extends(RouterContext, _super);
    function RouterContext(props, context) {
        var _this = _super.call(this, props, context) || this;
        if (process.env.NODE_ENV !== 'production') {
            if (!props.location || !props.matched) {
                throw new TypeError('"inferno-router" requires a "location" and "matched" props passed');
            }
        }
        return _this;
    }
    RouterContext.prototype.getChildContext = function () {
        return {
            router: this.props.router || {
                location: {
                    pathname: this.props.location,
                    baseUrl: this.props.baseUrl
                }
            }
        };
    };
    RouterContext.prototype.render = function (props) {
        return props.matched;
    };
    return RouterContext;
}(inferno_component_1.default));
exports.default = RouterContext;
